### DISCUSSION TITLE
About ColorDifference/DeltaE function return meaning

### DISCUSSION THREAD

**Discussion by explosion-mental** [Ogeon/palette#382] [Q&A] [answered]
tldl; Would be good to standardize the results from 1%-100%/0.1-1.0 or just add some documentation about the result types (for complete rookies c:) for color difference traits.

Hi, I've been using the [delta_e 1994](https://www.easyrgb.com/en/math.php) manually implemented and I've been guiding the result of such function with [this simple and easy table](https://zschuessler.github.io/DeltaE/learn).
However, when I switch to `palette` it seems that other methods, such as, `improved_difference` (meaning, CIEDE2000) and other functions/traits don't behave that way. Only `.distance()` works with the table I've linked before, and I'm not sure what to look for with other implementations. 
[This is very interesting](https://github.com/Ogeon/palette/blob/c54efbd43c03267713da337bd72005c9d0390598/palette/src/color_difference.rs#L99) as well, and If I understand correctly, it modifies the **result** of the standard function (from a quick glance at the doc) and I'm not sure how to handle that.

Background and use case: I use that as a configurable `threshold` in [`wallust`](https://codeberg.org/explosion-mental/wallust) in which I use the table mentioned to explain users how a `threshold` works (although is not a variable that many modify).

**Comment by Ogeon** [comment]
Hi, some sort of table like that could be a nice addition, where it's possible. They are currently implemented to be proportional to the scale of the source values, but it's not impossible that a bug or inconsistency made its way in. The CIEDE2000 implementation may be in need of a checkup, due to its age. Do you have any examples of when they differ and what you are comparing?

The traits with "improved" in their names are indeed tweaked variants of the original formulas. The idea with that research paper is to improve the average result by slightly modifying the output. You can ignore those if you prefer the original formulas.

Also, wallust looks really nice! 👀 Cool!

**Reply by explosion-mental** [reply]
> Do you have any examples of when they differ and what you are comparing?
```rust
fn main() {
    use palette::color_difference::Ciede2000;
    use palette::color_difference::ImprovedCiede2000;
    use palette::{IntoColor, Srgb};
    let rgb1 = Srgb::new(0.1, 0.2, 0.3);
    let rgb2 = Srgb::new(0.2, 0.1, 0.2);

    let lab1: palette::Lab = rgb1.into_color();
    let lab2: palette::Lab = rgb2.into_color();
    // println!("This is {lab1:?} and {lab2:?} \n\t=> DIFF: {}", lab1.difference(lab2));
    // println!("This is {lab1:?} and {lab2:?} \n\t=> DIFF2: {}", lab1.improved_difference(lab2));
    assert_eq!(lab1.difference(lab2), lab1.improved_difference(lab2));
}
```
I think accomoding the "threshold" in the same way done as the `.improved_differece` would be the _solution_? as in `color1.improved_difference(color2) <= (1.26*threshold.powf(0.55))` ¿ 

> They are currently implemented to be proportional to the scale of the source values, but it's not impossible that a bug or inconsistency made its way in.

I'm not sure what that means tbh.. And no, I don't think it's a bug, just lack of knowledge, I think adding some bit of documentation for it would be good.

> Also, wallust looks really nice! 👀 Cool!

Thanks! This is the [example in practice](https://codeberg.org/explosion-mental/wallust/src/commit/cb33ea7bd2f0d2ad188fe43ac0d5a4cca82ee9d4/src/colorspaces/mod.rs#L233)

Edit:
> The traits with "improved" in their names are indeed tweaked variants of the original formulas. The idea with that research paper is to improve the average result by slightly modifying the output. You can ignore those if you prefer the original formulas.

I did notice that, and yes, they produce "better" (much harmonic) color palettes!. Problem is that in my program the max threshold for the improved diff is ~8 (which produces a nice scheme), so maybe, like I said above, all I need is to also apply the `1.26*deltaE^0.55` to the threshold as well.

**Reply by Ogeon** [reply]
I think I see where the confusion lies. If I understand it correctly, you shouldn't necessarily need to apply the same formula, but you may still need to adjust the threshold. When you picked the number 8, you probably did that by looking at what looks good. That carries over a bias from the (in)accuracy of the difference formula. When you reduce the average bias by changing the formula, your threshold doesn't represent the perceived difference anymore. A bit like adapting your eye-hand coordination to a pair of glasses that distort your view.

If you apply the same formula as in the "improved" version, you warp your threshold space like they warp the distance space. I'm not sure if that will yield nice results. You could also do threshold selection by eye and it may or may not give you a better looking result. The "improved" scale is supposed to better represent what you see, after all, so I think it makes sense to not modify the threshold input.

I hope that makes sense. It's really cool that the adjustments give visible improvements, though!

**Reply by explosion-mental** [reply]
Wow.. Yes.
I can give you an example, and applying the formula does "solve" the issue.

![improved-1994](https://github.com/Ogeon/palette/assets/72536755/951c982c-d5e6-4780-bf73-cce00d76f137)

The left side is the ciede2000 improved and right is good old delta_e1994 (which is a simple euclidean equation irc) With the `threshold` being modified as `1.26* threshold.powf(0.55)`. While the results may not be that "pleasant" it does have fidelity with the prominent colors from the image.

> The "improved" scale is supposed to better represent what you see, after all, so I think it makes sense to not modify the threshold input.

Well I'm trying to generalize the table I linked in the last message in a way that if a user selects `20` it should "work out" for all the colorspaces given, as in it shouldn't require modification.

**Reply by Ogeon** [reply]
Hmm, maybe my phrasing wasn't the best. What I mean is that each formula changes the meaning of "20", or "8", or any other value along the scale. If you use a table for formula A, it's not necessarily right for formula B, so it will likely need to be adjusted between formulas. Ideally, a difference of 10 should look half as large as a difference of 20. If that's the case depends on how well the formula predicts the difference.

**Reply by explosion-mental** [reply]
Thats the point Im trying to understand. Im not sure how to tweak the threshold since I dont know what "scale" follows such operation, as the only ref I have is for good old deltae.

**Reply by Ogeon** [reply]
Ah, I think I get it now, sorry. Do you want to make it possible to choose both color difference formula and threshold value? While still having the same scale for the threshold for each combination? Or just to switch to a new formula, but keep the old scale?

I'm not sure that's easily possible. You can perhaps try to analyse the values and find a good enough conversion (like the one you already suggested), but some aspects of the formulas aren't directly comparable. Especially how the older formulas don't correctly take saturation into account, as the article you linked mentioned.

**Reply by Ogeon** [reply, marked as answer]
The table in the article is based on one person's subjective impressions, so it's not really based on a scientific process. If you run a number of color pairs through the algorithm you want to use and try to describe how different they are, you will have a similar table.

**Reply by explosion-mental** [reply]
> The table in the article is based on one person's subjective impressions

Right... Well this is embarrassing, there is no "scale" then. That is up to the observer(s).

Thanks for all of the clarifications :) 

I'll have to test and figure out wether to keep presenting the threshold to users to configure (meaning trying to keep it consistent) or hardcode one that has good enough results (and maybe iterate from a higher to lower threshold)

**Reply by Ogeon** [reply]
A table can probably be made, and it's probably going to look similar to that one for that formula. You may be able to find some way to have your own "meta scale", if you calibrate it to a number of differences. Like say that 100 is the difference between white and black or two other very different colors, 50 is some difference between, etc. Or take values from one formula (CIEDE2000 with or without the improvement, for example) and try to fit that curve to values from the other formulas.

I hope you find a nice solution and sorry for not really having a specific one.

**Reply by explosion-mental** [reply]
Forget all said, I've hesitated to use this lib because a knowledge gap, but the features and conveniences it gives are pretty good (although it doesn't have a complementary trait, which shouldn't be that hard to impl) sorry for any inconveniences and thanks for the crate :)

pd: apparently, you can use: `type Color: Mix<Scalar = f32>`, which is how I solved an issue.

**Reply by Ogeon** [reply]
Good that you found the solution! I wasn't fast enough. The color types can use different number types, so the traits need to be open for that as well.

Also, don't worry about asking questions. I'm happy to help if I can, and each question highlights a point where a knowledge gap could potentially be closed. The same for suggesting features. It makes it easier to know what to add next.
